src/machines/java/com/enderio/machines/common/menu/ImpulseHopperMenu.java

```diff
@@ -0,0 +1,74 @@
+package com.enderio.machines.common.menu;
+
+import org.apache.logging.log4j.LogManager;
+
+import com.enderio.machines.common.blockentity.ImpulseHopperBlockEntity;
+import com.enderio.machines.common.init.MachineMenus;
+
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.world.entity.player.Inventory;
+import net.minecraft.world.inventory.MenuType;
+import net.minecraft.world.level.block.entity.BlockEntity;
+
+public class ImpulseHopperMenu extends MachineMenu<ImpulseHopperBlockEntity>{
+
+
+    public ImpulseHopperMenu(ImpulseHopperBlockEntity blockEntity, Inventory inventory, int pContainerId) {
+        super(blockEntity, inventory, MachineMenus.IMPULSE_HOPPER.get(), pContainerId);
+        if (blockEntity != null) {
+            for(int j = 0; j < 2; ++j) {
+                for(int k = 0; k < 6; ++k) {
+                    this.addSlot(new MachineSlot(blockEntity.getInventory(), k + j * 6, 8 + 36 + k * 18, 9 + j * 54));
+                }
+            }
+            for(int k = 0; k < 6; ++k) {
+                this.addSlot(new GhostMachineSlot(blockEntity.getInventory(), 12 + k , 8 + 36 + k * 18, 9 + 27));
+            }
+            this.addSlot(new MachineSlot(blockEntity.getInventory(), 18, 11, 60));
+        }
+        addInventorySlots(8, 84);
+    }
+
+    public static ImpulseHopperMenu factory(@javax.annotation.Nullable MenuType<ImpulseHopperMenu> pMenuType, int pContainerId, Inventory inventory, FriendlyByteBuf buf) {
+        BlockEntity entity = inventory.player.level.getBlockEntity(buf.readBlockPos());
+        if (entity instanceof ImpulseHopperBlockEntity castBlockEntity)
+            return new ImpulseHopperMenu(castBlockEntity, inventory, pContainerId);
+        LogManager.getLogger().warn("couldn't find BlockEntity");
+        return new ImpulseHopperMenu(null, inventory, pContainerId);
+    }
+
+    //	//Fix dragging items still being removed when the slot is a ghost slot.
```

Comment:
What is this commented method left for?

Reply:
It's one of mine, but I think it was fixed with your slot impl? Or maybe not cause dragging logic is a pain.

Reply:
Dragging looks like it is fixed. Removing the commented code